parser/parser_system1.c

```diff
@@ -250,57 +280,54 @@ static int parse_disks(struct pv_state *s, char *value)
 	t = tokv + 1;
 	while ((str = pv_json_array_get_one_str(value, &size, &t))) {
 		struct pv_disk *d;
-		char *tmp;
 		jsmntok_t *diskv;
 		int diskc;
 
 		if (jsmnutil_parse_json(str, &diskv, &diskc) <= 0) {
 			pv_log(ERROR, "Invalid disk entry");
-			continue;
+			goto out;
 		}
 
 		d = pv_disk_add(s);
 		if (!d) {
+			pv_log(ERROR, "cannot add new disk");
 			free(diskv);
 			goto out;
 		}
 
 		d->name = pv_json_get_value(str, "name", diskv, diskc);
 		d->path = pv_json_get_value(str, "path", diskv, diskc);
-		d->options = pv_json_get_value(str, "options", diskv, diskc);
+		d->mount_target =
+			pv_json_get_value(str, "mount_target", diskv, diskc);
+		d->mount_ops =
+			pv_json_get_value(str, "mount_options", diskv, diskc);
+		d->format_ops =
+			pv_json_get_value(str, "format_options", diskv, diskc);
+		d->provision =
+			pv_json_get_value(str, "provision", diskv, diskc);
+		d->provision_ops = pv_json_get_value(str, "provision_options",
+						     diskv, diskc);
 		d->uuid = pv_json_get_value(str, "uuid", diskv, diskc);
+		d->type = get_type(str, diskv, diskc);
 
-		tmp = pv_json_get_value(str, "type", diskv, diskc);
-		if (!strcmp(tmp, "directory"))
-			d->type = DISK_DIR;
-		else if (!strcmp(tmp, "dm-crypt-versatile"))
-			d->type = DISK_DM_CRYPT_VERSATILE;
-		else if (!strcmp(tmp, "dm-crypt-caam"))
-			d->type = DISK_DM_CRYPT_CAAM;
-		else if (!strcmp(tmp, "dm-crypt-dcp"))
-			d->type = DISK_DM_CRYPT_DCP;
-		else {
-			d->type = DISK_UNKNOWN;
+		if (d->type == DISK_UNKNOWN) {
+			pv_log(ERROR, "cannot add new disk, type = UNKNOWN");
 			goto out;
 		}
 
-		free(tmp);
-		tmp = NULL;
+		d->format = get_format(str, diskv, diskc);
```

Comment:
I'd go to out in case this returns DISK_UNKNOWN

Reply:
Added in https://github.com/pantavisor/pantavisor/pull/363